### PR TITLE
feat(probabilistic_occupancy_grid_map): add time_keeper

### PR DESCRIPTION
**Pull request by a-maumau**
## Description
This PR will add time keeper feature to `probabilistic_occupancy_grid_map`.
For testing, use the argument `occupancy_grid_map_method:=` to specify the occupancy grid map method.

Need to merge with https://github.com/autowarefoundation/autoware_launch/pull/1138 to run in autoware.

Also, I fix the plugin name for `multi_lidar_pointcloud_based` case.

---
Sample output from autoware_debug_tools (updated from original comment)
```Text
- pointcloud_based
100.00% onPointcloudWithObstacleAndRaw: total 552.45 [ms], avg. 9.21 [ms], run count: 60
    ├── 84.75% create_occupancy_grid_map: total 468.20 [ms], avg. 7.80 [ms], run count: 60
    ├── 14.59% update_and_publish_occupancy_grid_map: total 80.62 [ms], avg. 1.34 [ms], run count: 60
    │   ├── 0.69% OccupancyGridMapToMsgPtr: total 3.79 [ms], avg. 0.06 [ms], run count: 60
    │   └── 13.91% rest: 76.83 [ms]
    └── 0.66% rest: 3.63 [ms]

- laserscan_based
100.00% onLaserscanPointCloud2WithObstacleAndRaw: total 335.46 [ms], avg. 6.58 [ms], run count: 51
    ├── 0.93% convertLaserscanToPointCLoud2: total 3.13 [ms], avg. 0.06 [ms], run count: 51
    ├── 34.94% height_filter: total 117.20 [ms], avg. 2.30 [ms], run count: 51
    ├── 19.86% transformPointcloud: total 66.63 [ms], avg. 1.31 [ms], run count: 51
    ├── 44.12% create_occupancy_grid_map: total 148.01 [ms], avg. 2.90 [ms], run count: 51
    │   ├── 19.65% update_and_publish_occupancy_grid_map: total 65.93 [ms], avg. 1.29 [ms], run count: 51
    │   │   ├── 0.93% OccupancyGridMapToMsgPtr: total 3.12 [ms], avg. 0.06 [ms], run count: 51
    │   │   └── 18.72% rest: 62.81 [ms]
    │   └── 24.47% rest: 82.08 [ms]
    └── 0.15% rest: 0.49 [ms]

- multi_lidar_pointcloud_based
100.00% onGridMap: total 855.03 [ms], avg. 2.57 [ms], run count: 333
    ├── 99.84% publish: total 853.70 [ms], avg. 7.69 [ms], run count: 111
    │   ├── 6.52% merge_grid_map: total 55.78 [ms], avg. 0.50 [ms], run count: 111
    │   │   ├── 6.00% OccupancyGridMsgToGridMap: total 51.30 [ms], avg. 0.15 [ms], run count: 333
    │   │   └── 0.52% rest: 4.48 [ms]
    │   ├── 69.21% SingleFrameOccupancyFusion: total 591.74 [ms], avg. 5.33 [ms], run count: 111
    │   │   ├── 68.98% create_fused_map: total 589.78 [ms], avg. 5.31 [ms], run count: 111
    │   │   │   ├── 68.58% cost_fusion_loop: total 586.35 [ms], avg. 5.28 [ms], run count: 111
    │   │   │   └── 0.40% rest: 3.42 [ms]
    │   │   └── 0.23% rest: 1.96 [ms]
    │   ├── 1.66% OccupancyGridMapToMsgPtr: total 14.17 [ms], avg. 0.06 [ms], run count: 222
    │   ├── 0.10% buffer_copy_loop: total 0.82 [ms], avg. 0.01 [ms], run count: 111
    │   └── 22.36% rest: 191.19 [ms]
    └── 0.16% rest: 1.34 [ms]
```

## Related links
https://github.com/autowarefoundation/autoware_launch/pull/1138

**Parent Issue:**

- Link

## How was this PR tested?
Checked with running autoware's sample-rosbag scene with all three options (`pointcloud_based`, `laserscan_based`, `multi_lidar_pointcloud_based`).

## Notes for reviewers

None.

## Interface changes

None.

## Effects on system behavior

None.

### CODE REVIEW

**Comment by a-maumau** [comment]
I added more timekeepers to make the processing time clearer (with less `rest` in the visualization) in cab3891 .

sample output:
```Text
- pointcloud_based
100.00% onPointcloudWithObstacleAndRaw: total 552.45 [ms], avg. 9.21 [ms], run count: 60
    ├── 84.75% create_occupancy_grid_map: total 468.20 [ms], avg. 7.80 [ms], run count: 60
    ├── 14.59% update_and_publish_occupancy_grid_map: total 80.62 [ms], avg. 1.34 [ms], run count: 60
    │   ├── 0.69% OccupancyGridMapToMsgPtr: total 3.79 [ms], avg. 0.06 [ms], run count: 60
    │   └── 13.91% rest: 76.83 [ms]
    └── 0.66% rest: 3.63 [ms]

- laserscan_based
100.00% onLaserscanPointCloud2WithObstacleAndRaw: total 335.46 [ms], avg. 6.58 [ms], run count: 51
    ├── 0.93% convertLaserscanToPointCLoud2: total 3.13 [ms], avg. 0.06 [ms], run count: 51
    ├── 34.94% height_filter: total 117.20 [ms], avg. 2.30 [ms], run count: 51
    ├── 19.86% transformPointcloud: total 66.63 [ms], avg. 1.31 [ms], run count: 51
    ├── 44.12% create_occupancy_grid_map: total 148.01 [ms], avg. 2.90 [ms], run count: 51
    │   ├── 19.65% update_and_publish_occupancy_grid_map: total 65.93 [ms], avg. 1.29 [ms], run count: 51
    │   │   ├── 0.93% OccupancyGridMapToMsgPtr: total 3.12 [ms], avg. 0.06 [ms], run count: 51
    │   │   └── 18.72% rest: 62.81 [ms]
    │   └── 24.47% rest: 82.08 [ms]
    └── 0.15% rest: 0.49 [ms]

- multi_lidar_pointcloud_based
100.00% onGridMap: total 855.03 [ms], avg. 2.57 [ms], run count: 333
    ├── 99.84% publish: total 853.70 [ms], avg. 7.69 [ms], run count: 111
    │   ├── 6.52% merge_grid_map: total 55.78 [ms], avg. 0.50 [ms], run count: 111
    │   │   ├── 6.00% OccupancyGridMsgToGridMap: total 51.30 [ms], avg. 0.15 [ms], run count: 333
    │   │   └── 0.52% rest: 4.48 [ms]
    │   ├── 69.21% SingleFrameOccupancyFusion: total 591.74 [ms], avg. 5.33 [ms], run count: 111
    │   │   ├── 68.98% create_fused_map: total 589.78 [ms], avg. 5.31 [ms], run count: 111
    │   │   │   ├── 68.58% cost_fusion_loop: total 586.35 [ms], avg. 5.28 [ms], run count: 111
    │   │   │   └── 0.40% rest: 3.42 [ms]
    │   │   └── 0.23% rest: 1.96 [ms]
    │   ├── 1.66% OccupancyGridMapToMsgPtr: total 14.17 [ms], avg. 0.06 [ms], run count: 222
    │   ├── 0.10% buffer_copy_loop: total 0.82 [ms], avg. 0.01 [ms], run count: 111
    │   └── 22.36% rest: 191.19 [ms]
    └── 0.16% rest: 1.34 [ms]
```

**Review by YoshiRi — COMMENTED** [review]
@a-maumau Thanks for you job!
Could you also add PR in autoware_launch to update config files in autoware_launch?
See https://github.com/autowarefoundation/autoware_launch/tree/main/autoware_launch/config/perception/occupancy_grid_map

**Comment by a-maumau** [comment]
@YoshiRi 

> Could you also add PR in autoware_launch to update config files in autoware_launch?

You mean this PR (it is in the top PR comment) ?
- https://github.com/autowarefoundation/autoware_launch/pull/1138

or I need to make other PR?

File: perception/autoware_probabilistic_occupancy_grid_map/src/fusion/synchronized_grid_map_fusion_node.cpp [hunk 1]
```diff
@@ -260,19 +274,28 @@ void GridMapFusionNode::timer_callback()
...
   // merge available gridmap
   std::vector<OccupancyGridMapFixedBlindSpot> subscribed_maps;
   std::vector<double> weights;
-  for (const auto & e : gridmap_dict_) {
-    if (e.second != nullptr) {
-      subscribed_maps.push_back(GridMapFusionNode::OccupancyGridMsgToGridMap(*e.second));
-      weights.push_back(input_topic_weights_map_[e.first]);
-      latest_stamp = e.second->header.stamp;
-      height = e.second->info.origin.position.z;
+  {  // scope for the timekeeper to track the time spent for merging grid map
```
Comment (technolojin) [inline perception/autoware_probabilistic_occupancy_grid_map/src/fusion/synchronized_grid_map_fusion_node.cpp:287]:
```suggestion
  {  // merging grid map
```

File: perception/autoware_probabilistic_occupancy_grid_map/src/fusion/synchronized_grid_map_fusion_node.cpp [hunk 2]
```diff
@@ -336,42 +362,58 @@ OccupancyGridMapFixedBlindSpot GridMapFusionNode::SingleFrameOccupancyFusion(
...
-    occupancy_grid_maps[0].getResolution());
-  fused_map.updateOrigin(
-    gridmap_origin.position.x - fused_map.getSizeInMetersX() / 2,
-    gridmap_origin.position.y - fused_map.getSizeInMetersY() / 2);
-
-  // fix origin of each map
-  for (auto & map : occupancy_grid_maps) {
-    map.updateOrigin(fused_map.getOriginX(), fused_map.getOriginY());
-  }
+  {  // add scope for time keeper
```
[the remark this answers is not in the document]
Comment (a-maumau) [reply perception/autoware_probabilistic_occupancy_grid_map/src/fusion/synchronized_grid_map_fusion_node.cpp]: this was already removed in [e97e3ec](https://github.com/autowarefoundation/autoware.universe/pull/8601/commits/e97e3ec4a6b5cb7c9c04f151b50a64a3519a72ff)

File: perception/autoware_probabilistic_occupancy_grid_map/src/fusion/synchronized_grid_map_fusion_node.cpp [hunk 3]
```diff
@@ -336,42 +362,58 @@ OccupancyGridMapFixedBlindSpot GridMapFusionNode::SingleFrameOccupancyFusion(
...
+    }
 
-  // assume map is same size and resolutions
-  for (unsigned int x = 0; x < fused_map.getSizeInCellsX(); x++) {
-    for (unsigned int y = 0; y < fused_map.getSizeInCellsY(); y++) {
-      // get cost of each map
-      std::vector<unsigned char> costs;
-      for (auto & map : occupancy_grid_maps) {
-        costs.push_back(map.getCost(x, y));
+    {  // add scope for time keeper (cost_fusion_loop)
```
[the remark this answers is not in the document]
Comment (a-maumau) [reply perception/autoware_probabilistic_occupancy_grid_map/src/fusion/synchronized_grid_map_fusion_node.cpp]: this was already removed in [e97e3ec](https://github.com/autowarefoundation/autoware.universe/pull/8601/commits/e97e3ec4a6b5cb7c9c04f151b50a64a3519a72ff)

File: perception/autoware_probabilistic_occupancy_grid_map/src/fusion/synchronized_grid_map_fusion_node.cpp (hunk 1 above)
Comment (a-maumau) [reply perception/autoware_probabilistic_occupancy_grid_map/src/fusion/synchronized_grid_map_fusion_node.cpp]: I applied the suggestion in 26eef9f.

File: perception/autoware_probabilistic_occupancy_grid_map/src/laserscan_based_occupancy_grid_map/laserscan_based_occupancy_grid_map_node.cpp [hunk 4]
```diff
@@ -163,51 +184,77 @@ void LaserscanBasedOccupancyGridMapNode::onLaserscanPointCloud2WithObstacleAndRa
...
-
-  if (enable_single_frame_mode_) {
-    // publish
-    occupancy_grid_map_pub_->publish(OccupancyGridMapToMsgPtr(
-      map_frame_, laserscan_pc_ptr->header.stamp, gridmap_origin.position.z,
-      single_frame_occupancy_grid_map));
-  } else {
-    // Update with bayes filter
-    occupancy_grid_map_updater_ptr_->update(single_frame_occupancy_grid_map);
+  {  // scope for the timekeeper to track the time spent for transformPointcloud
```
Comment (technolojin) [inline perception/autoware_probabilistic_occupancy_grid_map/src/laserscan_based_occupancy_grid_map/laserscan_based_occupancy_grid_map_node.cpp:188]:
```suggestion
  {  // transform pointclouds
```

**Review by technolojin — APPROVED** [review]
LGTM (with minor comment)

File: perception/autoware_probabilistic_occupancy_grid_map/src/laserscan_based_occupancy_grid_map/laserscan_based_occupancy_grid_map_node.cpp (hunk 4 above)
Comment (a-maumau) [reply perception/autoware_probabilistic_occupancy_grid_map/src/laserscan_based_occupancy_grid_map/laserscan_based_occupancy_grid_map_node.cpp]: I applied suggestion in 19b0915 .

**Review by YoshiRi — APPROVED** [review]
Thanks! LGTM

I confirmed the result of pointcloud-based method with Vtune. ( Roughly 50% seems to be consumed by ROS pub/sub system) 
image

**Comment by technolojin** [comment]
@a-maumau 
Can yo resolve the cppcheck?

Screenshot from 2024-09-09 14-46-19

**Comment by a-maumau** [comment]
@technolojin I fix the cppcheck error.
It passed in my local env., so it should pass now.

**Comment by YoshiRi** [comment]
Restarted CI. Please wait for a while.

**Pull request merged**